apps/opencs/view/render/terraintextureselection.cpp

```diff
@@ -0,0 +1,109 @@
+#include "terraintextureselection.hpp"
+
+#include <osg/Geode>
+#include <osg/Geometry>
+#include <osg/PositionAttitudeTransform>
+
+CSVRender::TerrainTextureSelection::TerrainTextureSelection(osg::Group* parentNode, const CSMWorld::CellCoordinates& coords, const ESM::Land& esmLand)
+    :TerrainSelection {coords, esmLand, parentNode}
+{
+    osg::ref_ptr<osg::Geometry> geometry {new osg::Geometry{}};
+
+    // Color
+    osg::ref_ptr<osg::Vec4Array> colors {new osg::Vec4Array{}};
+    colors->push_back(osg::Vec4f(0.f, 0.5f, 0.f, 1.f));
+
+    geometry->setColorArray(colors, osg::Array::Binding::BIND_OVERALL);
+
+    // Primitive
+    osg::ref_ptr<osg::DrawArrays> drawArrays {new osg::DrawArrays{osg::PrimitiveSet::Mode::LINES}};
+
+    geometry->addPrimitiveSet(drawArrays);
+    geometry->getOrCreateStateSet()->setMode(GL_LIGHTING, osg::StateAttribute::Values::OFF);
+
+    mGeode = new osg::Geode{};
+    mGeode->addDrawable(geometry);
+    getBaseNode()->addChild(mGeode);
+}
+
+void CSVRender::TerrainTextureSelection::addToSelection(osg::Vec3d worldPos)
+{
+    const TerrainPos localPos {toTextureCoords(worldPos)};
+
+    if (std::find(mSelection.begin(), mSelection.end(), localPos) == mSelection.end())
+        mSelection.push_back(localPos);
+}
+
+void CSVRender::TerrainTextureSelection::toggleSelection(osg::Vec3d worldPos)
+{
+    const TerrainPos localPos {toTextureCoords(worldPos)};
+    const auto iter = std::find(mSelection.begin(), mSelection.end(), localPos);
+
+    if (iter != mSelection.end()) {
+        mSelection.erase(iter);
+    }
+    else {
+        mSelection.push_back(localPos);
+    }
+}
+
+void CSVRender::TerrainTextureSelection::deselect()
+{
+    mSelection.clear();
+}
+
+void CSVRender::TerrainTextureSelection::update()
+{
+    const ESM::Land::LandData* landData {getLandData()};
+
+    if (!landData)
+        return;
+
+    const auto geometry = static_cast<const osg::Geometry*>(mGeode->getDrawable(0));
+    const osg::ref_ptr<osg::Geometry> newGeometry {
+        geometry
+        ? new osg::Geometry{*geometry, osg::CopyOp::SHALLOW_COPY}
```

Comment:
I'd say it's probably best to start simple and optimize only after a performance issue is noticed. Anyway, going by the name, a shallow copy would only copy the pointer, so if you go that route a new drawArrays needs to be created instead of accessing the old one.